keps/sig-scheduling/20190221-even-pods-spreading.md

```diff
@@ -0,0 +1,350 @@
+  * [Pros/Cons](#proscons)
+  * [Conclusion](#conclusion)
+  * [Test Plan](#test-plan)
+  * [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+## Terms
+
+- **Topology:** describe a series of worker nodes which belongs to the same
+  region/zone/rack/hostname/etc. In terms of Kubernetes, they're defined and
+  grouped by node labels.
+- **Affinity**: if not specified particularly, "Affinity" refers to
+  `NodeAffinity`, `PodAffinity` and `PodAntiAffinity`.
+- **CA**: Cluster Autoscaler.
+  [CA](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler)
+  is a tool that automatically adjusts the size of the Kubernetes cluster upon
+  specific conditions.
+
+## Summary
+
+`EvenPodsSpreading` feature gives users more fine-grained control on
+distribution of pods scheduling, so as to achieve better high availability and
+resource utilization.
+
+## Motivation
+
+In Kubernetes, "Affinity" related directives are aimed to control how pods are
+scheduled - more packed or more scattering. But right now only limited options
+are offered: for `PodAffinity`, infinite pods can be stacked onto qualifying
+topology domain(s); for `PodAntiAffinity`, pods are scheduled onto each single
+topology domain exclusively.
+
+This is not an ideal situation if users want to put pods evenly across different
+topology domains - for the sake of high availability or saving cost. And regular
+rolling upgrade or scaling out replicas can also be problematic. See more
+details in [user stories](#user-stories).
+
+> Even pods spreading is a long-discussed topic, but it's only feasible to
+> hammer out the design/implementation details after recent Affinity performance
+> improvements.
+
+### Goals
+
+- Even spreading is calculated among pods instead of apps API (such as
+  Deployment, ReplicaSet).
+- Even spreading can be either a predicate (hard requirement) or a priority
+  (soft requirement).
+- Even spreading _might_ be implemented on limited topologies in initial
+  version.
+
+### Non-Goals
+
+- Even spreading is NOT calculated on an application basis. In other words, it's
+  not _only_ applied within replicas of an application.
+- "Max number of pods per topology" is NOT a goal.
+- Scale-down on an application is not guaranteed to achieve even pods spreading
+  in initial implementation.
+
+## Proposal
+
+### User Stories
+
+#### Story 1 - NodeAffinity
+
+As an application developer, I want my application pods to be scheduled onto
+specific topology domains (via NodeAffinity with spec "app in [zone1, zone2,
+zone3]"), but I don't want them to be stacked too much on one topology. (see
+[#68981](https://github.com/kubernetes/kubernetes/issues/68981))
+
+#### Story 2 - PodAffinity
+
+As an application developer, I want my application pods to co-exist with
+particular pods in the same topology domain (via PodAffnity), and I want them to
+be deployed onto separate nodes (or sub domains) as even as possible.
```

Comment:
Its not clear how this use case is not handled by the current pod affinity feature ?

Reply:
PodAffinity defines which pods "attracts" which pods on which topology domain, but it's out of scope for it to promise an even pods distribution.

Reply:
thanks, the wording was not clear.  When you say even spreading, you mean equal pods per domain or per combination of domains right ? It would be good to create an example showing what that would mean in case of even spreading for one domain and even spreading across n domains

Reply:
Sure, thanks for the advice. I will add examples.